storage/localstore/migration_test.go

```diff
@@ -0,0 +1,287 @@
+// Copyright 2019 The Swarm Authors
+// This file is part of the Swarm library.
+//
+// The Swarm library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package localstore
+
+import (
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"testing"
+)
+
+func TestOneMigration(t *testing.T) {
+	DbSchemaCurrent = DbSchemaSanctuary
+	defer func(v []migration) { allDbSchemaMigrations = v }(allDbSchemaMigrations)
+
+	ran := false
+	shouldNotRun := false
+	allDbSchemaMigrations = []migration{
+		{name: DbSchemaSanctuary, migrationFunc: func(db *DB) error {
+			ran = true
+			return nil
+		}},
+		{name: DbSchemaDiwali, migrationFunc: func(db *DB) error {
+			shouldNotRun = true // this should not be executed
+			return nil
+		}},
+	}
+
+	dir, err := ioutil.TempDir("", "localstore-test")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(dir)
+	baseKey := make([]byte, 32)
+	if _, err := rand.Read(baseKey); err != nil {
+		t.Fatal(err)
+	}
+
+	// start the fresh localstore with the sanctuary schema name
+	db, err := New(dir, baseKey, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = db.Close()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	DbSchemaCurrent = DbSchemaDiwali
+
+	// start the existing localstore and expect the migration to run
+	db, err = New(dir, baseKey, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	schemaName, err := db.schemaName.Get()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if schemaName != DbSchemaDiwali {
+		t.Errorf("schema name mismatch. got '%s', want '%s'", schemaName, DbSchemaDiwali)
+	}
+
+	if !ran {
+		t.Errorf("expected migration did not run")
+	}
+
+	if shouldNotRun {
+		t.Errorf("migration ran but shouldnt have")
+	}
+
+	err = db.Close()
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func TestManyMigrations(t *testing.T) {
+	DbSchemaCurrent = DbSchemaSanctuary
```

Comment:
DbSchemaCurrent should be restored after the test just it is done with allDbSchemaMigrations.